docs/SPECIFICATION.md

```diff
@@ -0,0 +1,703 @@
+# Onyxia API Specification
+<p style="text-align: right;"> (back to <a href='./README.md'>Onyxia API Documentation</a>) <p>
+
+
+## Table of Contents
+1. [Terminology](#terminology)
+2. [Endpoints](#endpoints)
+3. [Data model](#data-model)
+
+## Terminology
+
+The following documentation will have mentions to the different packages, classes and concepts with the subsequent terminology.  
+
+The packages they will be referred as:
+- `onyxia-api`: refers to the package `fr.insee.onyxia.api` that exposes the endpoints and handles the requests, with source code under `onyxia-api/src/`.
+- `onyxia-model`: refers to the package `fr.insee.onyxia.model` that structures the request and response body into Java objects, with the source code under `onyxia-model/src/`.
+- `helm-wrapper`: refers to the package `io.github.inseefrlab.helmwrapper` that creates a Java interface to the machine's [Helm](https://helm.sh/) binary, with the source code under `helm-wrapper/src/` (to be moved to a different repository as a standalone package).
+
+Classes will be either referred with their fully qualified class name, e.g., the user controller in `onyxia-api` is handled by the class `fr.insee.onyxia.api.controller.api.user.UserController` , or a combination of their name and the package, e.g., `UserController` from `onyxia-api`. A reference to the source code file is also possible to appear if a class is mentioned.
+
+The concepts terminology is the following:
+- *Region*: the configuration block for the deployment of Onyxia, it was originally created when migrating Onyxia from Marathon (Mesos DC/OS) to Kubernetes.
+- *Service*: a package that can be installed in behalf of an user, so that the later can freely use it as his own.
+- *Task*: a component of the Service that runs the logic, i.e., translated to a Pod within Kuberenetes.
+- *Catalog*: a compilation of Services that are available for launching by Onyxia.
+- *Group*: the grouping of Services for a Project, which the underlying translation of the concept is a Kubernetes namespace where the services are launched.
+- *Project*: the logical construct of a group of users, with a one-to-one relationship with a Group, i.e., each project has its own Kubernetes namespace.
+- *Quota*: the request and limitation of resources per user (or group) namespace.
+
+
+## Endpoints 
+This endpoints specification is covered in the Open API Specification document accessible in the `/api/` endpoint. In this documentation we go into further detail with explanation of utility and backend processes.
+
+### Headers
+For authenticated requests the headers `ONYXIA-REGION` and `ONYXIA-PROJECT` should be set (besides the `Authorization` header). `ONYXIA-PROJECT` will inform the API if you are requesting information for the User or for one of the Projects he belongs to.
```

Comment:
Both of those headers are optional. There are default implemented for both :  
`ONYXIA_REGION` defaults to the first region in the API configuration.  
`ONYXIA-PROJECT` defaults to the user's personal project (which basically means no project).

Reply:
Added the default mention here 367000278529589fa06243a9bfefcc29544b3845 , I feel that a mention of the headers is necessary (will look for it to also be present in the OpenAPI Specification). Do you agree?

Reply:
Yes, I saw that you managed to find a way to document them using OpenAPI, that's great. Maybe someday we could find a way to define them once and for all globally or at least avoid repeating ourselves but I'm not sure if that's possible. Maybe we could try creating a meta-annotation like `@UsesProjectHeader` that would mutualize the documentation. Not sure if that would work tho